### PR TITLE
**MOVED CHANGES TO PR https://github.com/PubMatic-OpenWrap/prebid-server/pull/498**

### PR DESCRIPTION
**Pull request by pm-nilesh-chate**
**MOVED CHANGES TO PR https://github.com/PubMatic-OpenWrap/prebid-server/pull/498**

### CODE REVIEW

File: analytics/openwrap/helper.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+	// 	controller.LoggerRecord.SetUID(uidCookie.Value)
+	// }
+	// if testConfigApplied {
+	// 	controller.LoggerRecord.SetTestConfigApplied(1)
+	// }
+	return wl
+}
+
+// CreateLoggerRecordFromRequest creates logger and tracker records from bidRequest data
+func (wlog *WloggerRecord) CreateLoggerRecordFromRequest(uaFromHTTPReq string, ao *analytics.AuctionObject) {
```
Comment (Pubmatic-Dhruv-Sonone) [inline analytics/openwrap/helper.go:31]: any reason **uaFromHTTPReq** is input to function and not extracted internally from request?

File: analytics/openwrap/helper.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+	// if nil != ao.Request.Ext {
+	// 	ext, ok := ao.Request.Ext.(*openrtb.ExtRequest)
+	// 	if ok {
+	// 		wlog.logAdPodPercentage(ext.AdPod)
+	// 	}
+	// }
+}
+
+// Send method
+func Send(client http.Client, loggerURL string, wl *WloggerRecord, gdprEnabled int) error {
```
Comment (Pubmatic-Dhruv-Sonone) [inline analytics/openwrap/helper.go:114]: Can't we set gdpr as part of PrepareLoggerFromRequest ?

File: modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	Floors      *openrtb_ext.PriceFloorRules `json:"floors"`
+
+	Exp          *int             `json:"exp"`
+	Banner       *Banner          `json:"banner"`
+	Native       *Native          `json:"native"`
+	Video        *Video           `json:"video"`
+	App          *openrtb2.App    `json:"app"`
+	Device       *openrtb2.Device `json:"device"`
+	Transparency *Transparency    `json:"transparency,omitempty"`
+	Regex        *bool            `json:"regex"`
```
Comment (pm-aadit-patil) [inline modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go:62]: Can we check if we use regex in the AdUnitCOnfig level. In adunit we have only regex supported at Root level

File: modules/pubmatic/openwrap/database/mysql/adunit_config.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	}
+	adunitConfigQuery = strings.Replace(adunitConfigQuery, profileIdKey, strconv.Itoa(profileID), -1)
+	adunitConfigQuery = strings.Replace(adunitConfigQuery, displayVersionKey, strconv.Itoa(displayVersionID), -1)
+	err := db.conn.QueryRow(adunitConfigQuery).Scan(&adunitConfigJSON)
+	if err != nil {
+		err = fmt.Errorf("[QUERY_FAILED] Name:[%v] Error:[%v]", "GetAdunitConfig", err.Error())
+		return nil, err
+	}
+
+	err = json.Unmarshal([]byte(adunitConfigJSON), &adunitConfig)
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/database/mysql/adunit_config.go:31]: * Should be abort the request on unmarshal failure?
* Verify behaviour of all Cache/DB calls is as is and use `error` only for log

File: modules/pubmatic/openwrap/database/database.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+	"github.com/prebid/prebid-server/modules/pubmatic/openwrap/models"
+	"github.com/prebid/prebid-server/modules/pubmatic/openwrap/models/adunitconfig"
+)
+
+type Database interface {
+	GetAdunitConfig(profileID, displayVersionID int) (*adunitconfig.AdUnitConfig, error)
+	GetActivePartnerConfigurations(pubId, profileId, displayVersion int) (map[int]map[string]string, error)
+	GetPubmaticSlotMappings(pubId int) map[string]models.SlotMapping
+	GetPublisherSlotNameHash(pubID int) map[string]string
+	GetWrapperSlotMappings(partnerConfigMap map[int]map[string]string, profileId, displayVersion int) map[int][]models.SlotMapping
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/database/database.go:13]: return error

File: modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import (
+	"encoding/json"
+
+	"github.com/prebid/openrtb/v17/openrtb2"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+// AdUnitConfig type definition for Ad Unit config parsed from stored config JSON
+type AdUnitConfig struct {
+	ConfigPattern string               `json:"configPattern"`
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go:12]: standardise th key names, case, etc.

File: modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	Content Content `json:"content,omitempty"`
+}
+
+type BannerConfig struct {
+	openrtb2.Banner
+	ClientConfig json.RawMessage `json:"clientconfig"`
+}
+
+type Banner struct {
+	Enabled *bool         `json:"enabled"`
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go:32]: can we remove pointer

File: modules/pubmatic/openwrap/models/utils.go [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+	// NYC_TODO: fix this temporary change
+}
+
+func GetRequestExtWrapper(request []byte, wrapperLocation ...string) (RequestExtWrapper, error) {
+	extWrapper := RequestExtWrapper{
+		SSAuctionFlag: -1,
+	}
+
+	if len(wrapperLocation) == 0 {
+		wrapperLocation = []string{"ext", "prebid", "bidderparams", "pubmatic", "wrapper"}
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/models/utils.go:28]: How can we use this location when pubmatic is throttled?. How does web request work

File: router/router_ow.go [hunk 9]
```diff
@@ -1,174 +1,37 @@
...
-
-// GetUIDSWrapper Openwrap wrapper method for calling /getuids endpoint
-func GetUIDSWrapper(w http.ResponseWriter, r *http.Request) {
-	getUID := endpoints.NewGetUIDsEndpoint(g_cfg.HostCookie)
-	getUID(w, r, nil)
-}
+// Support legacy APIs for a grace period.
+// not implementing middleware to avoid duplicate processing like read, unmarshal, write, etc.
+// handling the temporary middleware stuff in EntryPoint hook.
+func (r *Router) registerOpenWrapEndpoints(openrtbEndpoint, ampEndpoint httprouter.Handle) {
```
Comment (pm-nilesh-chate) [inline router/router_ow.go:20]: Add automation check for api path validation (trailing /, get, post, etc)

File: router/router.go
Comment (pm-nilesh-chate) [inline router/router.go:1]: check if we need to expose PBS endpoints like /bidders, etc

File: config/config.go [hunk 10]
```diff
@@ -485,9 +485,9 @@ type LMT struct {
 }
 
 type Analytics struct {
-	File     FileLogs `mapstructure:"file"`
-	Pubstack Pubstack `mapstructure:"pubstack"`
-	PubMatic PubMatic `mapstructure:"pubmatic"`
+	File     FileLogs   `mapstructure:"file"`
+	Pubstack Pubstack   `mapstructure:"pubstack"`
+	PubMatic PubMaticWL `mapstructure:"pubmatic"`
```
Comment (pm-nilesh-chate) [inline config/config.go:490]: Keep the naming convention consistent for analytics and its config. (OpenWrap vs PubMaticWL)

File: config/config.go [hunk 11]
```diff
@@ -521,8 +521,15 @@ type Pubstack struct {
 	ConfRefresh string         `mapstructure:"configuration_refresh_delay"`
 }
 
-type PubMatic struct {
-	Enabled bool `mapstructure:"enabled"`
+type PubMaticWL struct {
```
Comment (pm-nilesh-chate) [inline config/config.go:524]: move this to `_ow.go` file

File: endpoints/openrtb2/auction.go [hunk 12]
```diff
@@ -284,6 +289,11 @@ func rejectAuctionRequest(
 		response.ID = request.ID
 	}
 
+	// TODO merge this with success case
+	stageOutcomes := hookExecutor.GetOutcomes()
+	ao.HookExecutionOutcome = stageOutcomes
+	UpdateResponseExtOW(response, ao)
+
```
Comment (pm-nilesh-chate) [inline endpoints/openrtb2/auction.go:296]: review and club it

File: exchange/floors.go [hunk 13]
```diff
@@ -128,6 +128,11 @@ func enforceFloorToBids(bidRequestWrapper *openrtb_ext.RequestWrapper, seatBids
 
 			bidPrice := rate * bid.Bid.Price
 			if reqImp.BidFloor > bidPrice {
+				if bid.BidFloors != nil {
+					// need USD for analytics
+					// TODO: Move this to impCtx and use it in all places
+					bid.BidFloors.FloorValueUSD = getOriginalBidCpmUsd(reqImp.BidFloor, reqImpCur, conversions)
+				}
```
Comment (pm-nilesh-chate) [inline exchange/floors.go:135]: do currency conversion in logger

File: go.mod [hunk 14]
```diff
@@ -1,4 +1,4 @@
-module github.com/PubMatic-OpenWrap/prebid-server
```
Comment (pm-nilesh-chate) [inline go.mod:1]: revert this line

File: floors/enforce.go [hunk 15]
```diff
@@ -138,6 +138,11 @@ func enforceFloorToBids(bidRequestWrapper *openrtb_ext.RequestWrapper, seatBids
 
 				bidPrice := rate * bid.Bid.Price
 				if reqImp.BidFloor > bidPrice {
+					if bid.BidFloors != nil {
+						// Need USD for OW analytics
+						// TODO: Move this to better place where 'conversions' (deduced from host+request) is available
+						bid.BidFloors.FloorValueUSD = getOriginalBidCpmUsd(reqImp.BidFloor, reqImp.BidFloorCur, conversions)
+					}
```
Comment (pm-nilesh-chate) [inline floors/enforce.go:145]: clear this filed in response hook. for reject bid, add field in pbsortbseat

**Comment by pm-nilesh-chate** [comment]
**MOVED CHANGES TO PR https://github.com/PubMatic-OpenWrap/prebid-server/pull/498**

**Pull request closed without merging**